### PR TITLE
feat: add Discord and Twitch as signin providers

### LINKED ISSUE #92: Request: Support other popular social logins
What's missing on popular ones:
- Discord
- Twitter
- Apple

In addition to these currently supported on nhost.io:

Screen Shot 2022-02-03 at 7 11 15 PM

### PR DESCRIPTION
**Pull request by elitan**
closes #92

### CODE REVIEW

**Review by plmercereau — CHANGES_REQUESTED** [review]
`migrations/00003_discord_twitch_providers.sql`
rename migration name to use kebab-case instead of snake_case:
`migrations/00003_discord-twitch-providers.sql`

File: src/config/providers.ts [hunk 1]
```diff
@@ -232,6 +232,25 @@ const PROVIDERS = {
           },
         };
   },
+
+  get discord() {
+    return !castBooleanEnv('AUTH_PROVIDER_DISCORD_ENABLED')
+      ? null
+      : {
+          get clientID() {
+            return process.env.AUTH_PROVIDER_DISCORD_CLIENT_ID || '';
```
Comment (plmercereau) [inline src/config/providers.ts:241]: minor: use `??` instead of `||`

File: src/config/providers.ts [hunk 2]
```diff
@@ -232,6 +232,25 @@ const PROVIDERS = {
...
+          },
+          get clientSecret() {
+            return process.env.AUTH_PROVIDER_DISCORD_CLIENT_SECRET || '';
```
Comment (plmercereau) [inline src/config/providers.ts:244]: ditto

File: 0001-create-schema.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+-- https://github.com/hasura/graphql-engine/issues/3657
+CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;
+CREATE EXTENSION IF NOT EXISTS citext WITH SCHEMA public;
+CREATE OR REPLACE FUNCTION public.set_current_timestamp_updated_at() RETURNS trigger LANGUAGE plpgsql AS $$
+declare _new record;
+begin _new := new;
+_new."updated_at" = now();
+return _new;
+end;
+$$;
```
Comment (elitan) [inline 0001-create-schema.sql:12]: I don't understand why this eas added as a new file in this PR 🤔

Comment (plmercereau) [reply 0001-create-schema.sql:12]: It's in the repo's root

**Review by plmercereau — CHANGES_REQUESTED** [review]

**Comment by elitan** [comment]
🤦

**Review by plmercereau — APPROVED** [review]

File: src/config/providers.ts [hunk 4]
```diff
@@ -232,6 +232,43 @@ const PROVIDERS = {
...
+          get clientID() {
+            return process.env.AUTH_PROVIDER_DISCORD_CLIENT_ID ?? '';
+          },
+          get clientSecret() {
+            return process.env.AUTH_PROVIDER_DISCORD_CLIENT_SECRET ?? '';
+          },
+          get scope() {
+            return castStringArrayEnv('AUTH_PROVIDER_DISCORD_SCOPE', [
+              'identify',
+              'email',
```
Comment (fungilation) [inline src/config/providers.ts:249]: 👍🏼

File: src/config/providers.ts [hunk 5]
```diff
@@ -232,6 +232,43 @@ const PROVIDERS = {
...
+      : {
+          get clientID() {
+            return process.env.AUTH_PROVIDER_TWITCH_CLIENT_ID ?? '';
+          },
+          get clientSecret() {
+            return process.env.AUTH_PROVIDER_TWITCH_CLIENT_SECRET ?? '';
+          },
+          get scope() {
+            return castStringArrayEnv('AUTH_PROVIDER_TWITCH_SCOPE', [
+              'user:read:email',
```
Comment (fungilation) [inline src/config/providers.ts:267]: 👍🏼

**Review by fungilation — APPROVED** [review]
email scopes with client_id, client_secret for both discord and twitch providers. Great!

File: src/config/providers.ts (hunk 4 above)
Comment (elitan) [reply src/config/providers.ts:249]: Yep. Bot tested and working.

File: src/config/providers.ts (hunk 5 above)
Comment (elitan) [reply src/config/providers.ts:267]: Yep. Both tested and working.

**Pull request merged**